### PR TITLE
Christian Pinna's pull request 

### PR DESCRIPTION
**Pull request by ChrisPinna**
# Christian Pinna
Please write your full name here to make it easier to find your pull request.

# User stories 

Please list which user stories you've implemented (delete the ones that don't apply).

- [x] User story 1: "I want to instruct a plane to land at an airport"
- [x] User story 2: "I want to instruct a plane to take off from an airport and confirm that it is no longer in the airport"
- [x] User story 3: "I want to prevent landing when the airport is full"
- [x] User story 4: "I would like a default airport capacity that can be overridden as appropriate"
- [x] User story 5: "I want to prevent takeoff when weather is stormy"
- [x] User story 6: "I want to prevent landing when weather is stormy"

# README checklist

Does your README contains instructions for

- [ ] how to install,
- [ ] how to run,
- [ ] and how to test your code?

[Here is a pill](https://github.com/makersacademy/course/blob/main/pills/readmes.md) that can help you write a great README!

### CODE REVIEW

File: lib/plane.rb [hunk 1]
```diff
@@ -0,0 +1,26 @@
...
+
+class Plane
+
+  def initialize
+    @current_airport = nil
+    @flying = true
+  end
+
+  def land(airport)
+    raise 'error, plane is not flying' if @flying == false
```
Comment (adamwoodcock98) [inline lib/plane.rb:11]: Nice job bro - managed to include some of those edge cases, well done!

File: lib/plane.rb [hunk 2]
```diff
@@ -0,0 +1,26 @@
+require_relative './airport'
+
+class Plane
+
+  def initialize
+    @current_airport = nil
+    @flying = true
+  end
+
+  def land(airport)
```
Comment (adamwoodcock98) [inline lib/plane.rb:10]:
```suggestion
  def land_at(airport)
```
Using `land_at(airport)` would make this read like a full sentence

File: spec/airport_spec.rb [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+describe Airport do
+	it 'can creat a new airport' do
+			expect { Airport.new }.to_not raise_error
+	end
+
+	it 'can creat a new plane' do 
+			expect { Plane.new }.to_not raise_error
+	end
+
+	it 'plane object responds to land method' do
```
Comment (adamwoodcock98) [inline spec/airport_spec.rb:12]: Maybe create a `plane_spec.rb` file for these kinds of tests.

File: spec/airport_spec.rb [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+			10.times { airport.hanger << Plane.new }
+			plane = Plane.new
+			expect { plane.land(airport) }.to raise_error 'airport full'
+	end
+
+	it 'method storm? should return true if stormy' do
+			airport = Airport.new
+			airport.storm = true
+			expect(airport.storm).to eq(true)
+	end
```
Comment (adamwoodcock98) [inline spec/airport_spec.rb:55]: We need to avoid vacuous tests (from the rubic). In this one, with `airport.storm = true` you are setting the conditions yourself, so in the `expect(airport.storm).to eq(true)` section, regardless of your code implementation in the `airport.rb` class file, this test will be successful. I hope I'm explaining this okay. This is where mocks and doubles come in.

What you would be looking for here would be code to test the actual _behaviour_ of the class. something like:
```ruby
airport = Airport.new # in real practice, this would be a double, because your logic to stop a plane taking off is in the plane class file, something like:
airport = double("airport", :storm? => true)
expect { plane.take_off(airport) }.to raise_error 'take-off denied, there is a storm!'
```
Now you're using the behaviour of your code to test the state of the storm, because if there is a storm, your code should fail. I'll try and explain this more if you need me to.

File: spec/airport_spec.rb [hunk 5]
```diff
@@ -0,0 +1,84 @@
...
+			airport.storm = true
+			expect { plane.take_off(airport) }.to raise_error('take-off denied, there is a storm!')
+	end
+
+	it 'a plane should not be able to land if the weather at the airport is stormy' do	
+			airport = Airport.new
+			plane = Plane.new
+			airport.storm = true
+			expect { plane.land(airport) }.to raise_error('landing denied, there is a storm!')
+	end
```
Comment (adamwoodcock98) [inline spec/airport_spec.rb:70]: I've now seen that you've done what I put in that previous long comment, so that would change my advise to the fact that you could just delete that test, as through these two you're by default testing that the `storm` predicate works.

File: spec/airport_spec.rb [hunk 6]
```diff
@@ -0,0 +1,84 @@
...
+
+	it 'a plane should not be able to take-off from an airport it is not in' do
+		airport = Airport.new
+		plane = Plane.new
+		expect { plane.take_off(airport) }.to raise_error('plane is not in the specified airport')
+	end
```
Comment (adamwoodcock98) [inline spec/airport_spec.rb:76]: Nice!!!!

File: lib/plane.rb [hunk 7]
```diff
@@ -0,0 +1,26 @@
...
+    @current_airport = airport
+    @flying = false
+    airport.hanger << self
+  end
+
+  def take_off(airport)
+    raise 'plane is not in the specified airport' if airport != @current_airport
+    raise 'take-off denied, there is a storm!' if airport.storm == true
+    @flying = true
+    airport.hanger.reject! { |plane| plane == self }
```
Comment (adamwoodcock98) [inline lib/plane.rb:23]:
```suggestion
    airport.hanger.delete(self) # this would return the deleted plane
    airport.hanger -= [self] # this would return the array, without the deleted plane
```
I'm not 100% sure these would work with your code, I'm only suggesting because this line took me a little while to figure out, so just trying to maybe suggest something that improves readability - maybe you tried these suggestions and they didn't work, not trying to teach you to suck eggs!

**Review by adamwoodcock98 — COMMENTED** [review]
I think it's great how you've covered most of the edge cases, I was no where near that point so I'm impressed. My biggest takeaway was that the tests could definitely use some separation and use of mocks, but I know from the peer group you want a better understanding of this, so that makes sense. There's some better instructions in steps 19 & 20 of the Boris Bikes challenge.
But really well done Chris!
